# ![](/images/icons/help.gif)Using Bioconductor for Sequence Data

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("knitr")
options(width=100)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE)
```

```{r, echo=FALSE}
suppressPackageStartupMessages({
   library(GenomicRanges)
   library(GenomicAlignments) 
   library(Biostrings)
   library(Rsamtools)
   library(ShortRead)
   library(BiocParallel)
   library(rtracklayer)
   library(VariantAnnotation)
   library(AnnotationHub)
   library(BSgenome.Hsapiens.UCSC.hg19)
   library(RNAseqData.HNRNPC.bam.chr14)
})
ah = AnnotationHub()
```
_Bioconductor_ allows one to do analysis and comprehension of high-
throughput genomic data. We have a vast number of packages that allow
rigorous statistical analysis of large data while keeping technological 
artifacts in mind. We can also comprehend and visualize these results
in a biological context and reproduce these analysis easily. Different
types of analysis can be carried out using _Bioconductor_, for example 

- Sequencing : RNASeq, ChIPSeq, variants, copy number.. 
- Microarrays: expression, SNP, ...
- Domain specific analysis : Flow cytometry, Proteomics ..
 
For all these analysis, one has to import and work with diverse 
sequence-related file types, including fasta, fastq, BAM, gff, 
bed, and wig files, among others. _Bioconductor_ Packages 
support import, common and advanced sequence 
manipulation operations such as trimming, transformation, 
and alignment including quality assessment. 

<h2 id="contents"> Contents </h2>

* [Sequencing Resources](#pkgs)
* [Ranges Infrastructure](#range)
* [DNA /amino acid sequence from FASTA files](#fasta)
* [Reads from FASTQ files](#fastq)
* [Aligned Reads from BAM files](#bam)
* [Called Variants from VCF files](#vcf)
* [Genome Annotations from BED, WIG, GTF etc files](#rtracklayer)

<h2 id="pkgs">Sequencing Resources</h2>

Here is a illustrative description elaborating the different
file types at various stages in a typical analysis with the 
package names (in pink boxes) that one will use for each stage. 

![](sequencepkg.png)

The following packages illustrate the diversity of functionality
available; all are in the release version of _Bioconductor_.

* `r Biocpkg("IRanges")`, `r Biocpkg("GenomicRanges")` and
  `r Biocpkg("genomeIntervals")`
  for range-based (e.g., chromosomal regions) calculation, data
  manipulation, and general-purpose data
  representation. `r Biocpkg("Biostrings")` for alignment, pattern matching
  (e.g., primer removal), and data manipulation of large biological
  sequences or sets of sequences.

* `r Biocpkg("ShortRead")` and `r Biocpkg("Rsamtools")` for file I/O, 
  quality assessment,   and high-level, general purpose data summary.  
  `r Biocpkg("rtracklayer")` for   import and export of tracks on 
  the UCSC genome browser.

* `r Biocpkg("BSgenome")` for accessing and manipulating curated whole-genome
  representations.  `r Biocpkg("GenomicFeatures")` for annotation of sequence
  features across common genomes, `r Biocpkg("biomaRt")` for access to Biomart
  databases.

* `r Biocpkg("SRAdb")`  for querying and retrieving data from the 
  Sequence Read Archive.

* ChIP-seq and related (e.g., motif discovery, identification of
  high-coverage segments) activities are facilitated by packages such
  as
  `r Biocpkg("CSAR")`,   `r Biocpkg("chipseq")`,
  `r Biocpkg("ChIPseqR")`,   `r Biocpkg("ChIPsim")`,
  `r Biocpkg("ChIPpeakAnno")`,   `r Biocpkg("DiffBind")`,
  `r Biocpkg("iSeq")`,  `r Biocpkg("rGADEM")`,
  `r Biocpkg("segmentSeq")`, `r Biocpkg("BayesPeak")`,
  `r Biocpkg("PICS")`.

* Differential expression and RNA-seq style analysis can be
  accomplished with `r Biocpkg("edgeR")`, `r Biocpkg("DESeq")`,
  `r Biocpkg("baySeq")`, `r Biocpkg("DEGseq")`,
  `r Biocpkg("Genominator")`. The
  `r Biocpkg("DEXSeq")` package, available with R >= 2.14, allows exon-level
  differential representation.

Additional packages are classified in the 
[biocViews](/packages/release/BiocViews.html) hierarchy,
for instance under Software -> AssayTechnologies ->
HighThroughputSequencing.

<h2 id="range">Ranges Infrastructure</i></h2>

These packages rely very heavily on the *IRanges* / *GenomicRanges*
infrastructure. Thus we will begin with a quick introduction to
these and then cover different file types. 

The `r Biocpkg("GenomicRanges")` package allows us to associate a range 
with a sequence name (e.g., chromosome) and a strand.
GRanges is a vector of genomic locations and associated annotations. 
Each element in the vector is comprised of a sequence name, 
an interval, a strand, and optional metadata columns 
(e.g. score, GC content, etc.). 

```[r}
library(GenomicRanges)
GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),
      IRanges(1:10, width=5), strand='-',
      score=101:110, GC = runif(10))
```

Use help() to list the help pages in the `r Biocpkg("GenomicRanges")` package, 
and vignettes() to view and access available vignettes.

```{r eval=FALSE}
help(package="GenomicRanges")
vignette(package="GenomicRanges")
```

<h2 id="fasta">DNA /amino acid sequence from FASTA files</h2>

`r Biocpkg("Biostrings")` classes (XStringset eg: DNAString) are used to represent
DNA or amino acid sequences. In the example below we will construct a
DNAString and show some manipulations. 

```{r message=FALSE}
library(Biostrings)
d <- DNAString("TTGAAAA-CTC-N")
length(d)  #no of letters in the DNAString
```

We will download the file 'Homo_sapiens.GRCh38.cdna.all.fa' 
from Ensembl using `r Biocpkg("AnnotationHub")`.

```{r eval=FALSE}
library(AnnotationHub)
ah = AnnotationHub()
```
This file is downloaded as a FaFile which can be read in using
readFasta from the `r Biocpkg("ShortRead")` package

```{r}
ah2 <- query(ah, c("fasta", "homo sapiens", "Ensembl"))
fl <- ah2[["AH18522"]]
fl
```
We will open the file and create an index from the fasta 
file using `r Biocpkg("Rsamtools")`.
```{r}
library(Rsamtools)
fa <- open(fl)
idx <- scanFaIndex(fa)
idx
```
scanFaIndex has read in the sequence names and and widths of 
recorded in an indexed fasta file, returning the information 
as a GRanges object. scanFa returns the sequences indicated by 
param as a DNAStringSet instance.

```{r}
dna <- scanFa(fa, param=idx[1:2])
dna
```
`r Biocpkg("BSgenome")` packages inside _Bioconductor_ contain whole genome
sequences as distributed by ENSEMBL, NCBI and others.
In this next example we will load the whole genome sequence 
for Homo sapiens from UCSC's `hg19` build and calculate the GC 
content across chromosome 14. 

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)

chr14_range = GRanges("chr14", IRanges(1, seqlengths(Hsapiens)["chr14"]))
chr14_dna <- getSeq(Hsapiens, chr14_range)
letterFrequency(chr14_dna, "GC", as.prob=TRUE)
```


<h2 id="fastq"> Reads from FASTQ files</h2>

`r Biocpkg("ShortRead")` package from _Bioconductor_ can be used for summarizing 
quality of fastq files from within R. Here we illustrate a quick
example where one can read in multiple fasta files, collect some 
statistics and generate a report about the same. 

`r Biocpkg("BiocParallel")` is another package from _Bioconductor_ which 
parallelizes this task and speeds up the process. 

```{r eval=FALSE}
## 1. attach ShortRead and BiocParallel
library(ShortRead)
library(BiocParallel)

## 2. create a vector of file paths
fls <- dir("~/fastq", pattern="*fastq", full=TRUE)

## 3. collect statistics
stats0 <- qa(fls)

## 4. generate and browse the report
if (interactive())
    browseURL(report(stats0))
```

<h2 id="bam">Aligned Reads from BAM files</h2>

The `r Biocpkg("GenomicAlignments")` package is used to input reads aligned 
to a reference genome. 

In this next example, we will read in a BAM file 
and specifically read in reads 
supporting the junction identified at position 
19653707 + 66M = 19653773 of chromosome 14.

The package `r Biocexptpkg("RNAseqData.HNRNPC.bam.chr14_BAMFILES")`
contains 8 BAM files. We will use only the 
first BAM file. We will load the software packages
and the data package, construct a GRanges
with our region of interest and use summarizeJunctions
to find reads in our region of interest. 

```{r}
## 1. load software packages
library(GenomicRanges)
library(GenomicAlignments)

## 2. load sample data
library('RNAseqData.HNRNPC.bam.chr14')
bf <- BamFile(RNAseqData.HNRNPC.bam.chr14_BAMFILES[[1]], asMates=TRUE)

## 3. define our 'region of interest'
roi <- GRanges("chr14", IRanges(19653773, width=1)) 

## 4. alignments, junctions, overlapping our roi
paln <- readGAlignmentsList(bf)
j <- summarizeJunctions(paln, with.revmap=TRUE)
j_overlap <- j[j %over% roi]

## 5. supporting reads
paln[j_overlap$revmap[[1]]]
```
For a detailed tutorial on working with BAM files
do check out this detailed [vignette of GenomicAlignments](http://bioconductor.org/packages/devel/bioc/vignettes/GenomicAlignments/inst/doc/OverlapEncodings.pdf)


<h2 id="vcf">Called Variants from VCF files</h2>

VCF text files contain meta-information lines, a header 
line with column names, data lines with information about 
a position in the genome, and optional genotype information 
on samples for each position. 


Data are parsed into a VCF object with readVcf
from `r Biocpkg("VariantAnnoation")`

```{r}
library(VariantAnnotation)
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- readVcf(fl, "hg19")
```

An excellent workflow on working with Variants can be found 
[here](http://bioconductor.org/help/workflows/variants/)


<h2 id="rtracklayer">Genome Annotations from BED, WIG, GTF etc files</h2>

`r Biocpkg("rtracklayer")` import and export 
functions can read in many common 
file types, e.g., BED, WIG, GTF, â€¦, in addition to querying 
and navigating the UCSC genome browser.

`r Biocpkg("rtracklayer")` contains a 'test' BED file which we will read in here

```{r}
library(rtracklayer)
test_path <- system.file("tests", package = "rtracklayer")
test_bed <- file.path(test_path, "test.bed")
  
test <- import(test_bed, format = "bed")
test
```

The file is read in and returned as a GRanges to the user.
A more detailed tutorial can be found [here](http://bioconductor.org/packages/devel/bioc/vignettes/rtracklayer/inst/doc/rtracklayer.pdf) 

`r Biocpkg("AnnotationHub")` also contains a variety of genomic annotation files
(eg BED, GTF, BigWig) which use import() from rtracklayer 
behind the scenes. For a detailed tutorial the user is referred to
[Annotation workflow](http://bioconductor.org/help/workflows/annotation/annotation/)
and [AnnotationHub HOW TO vignette](http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html)


```{r}
sessionInfo()
```

